components/eam/src/physics/cam/ndrop.F90

```diff
@@ -1339,270 +1312,73 @@ subroutine activate_modal(wbar, sigw, wdiab, wminf, wmaxf, tair, rhoair,  &
+! BJG g renamed to gthermfac, defined here since same for all modes in implementation
+   gthermfac=1._r8/(rhoh2o/(diff0*rhoair*qs)                                    &
+           +latvap*rhoh2o/(conduct0*tair)*(latvap/(rh2o*tair)-1._r8))
+! BJG wdiab = 0 whenever subroutine is called, so removed from argument list
+!      wnuc=wbar+wdiab
+   wnuc = wbar
+!  BJG  below is always true if this point is reached.
+!      if(wnuc.gt.0._r8)then
+   alw=alpha*wnuc
+   etafactor1=alw*sqrt(alw)
+
+!  BJG  Here compute smc, zeta, eta for all modes for maxsat calculation
+
+   do imode=1,nmode
+     if(volume(imode) > 1.e-39_r8 .and. na(imode) > 1.e-39_r8)then
          !            number mode radius (m)
          !           write(iulog,*)'alogsig,volc,na=',alogsig(m),volc(m),na(m)
-         amcube(m)=(3._r8*volume(m)/(4._r8*pi*exp45logsig(m)*na(m)))  ! only if variable size dist
+         amcube(imode)=(3._r8*volume(imode)/(4._r8*pi*exp45logsig(imode)*na(imode)))  ! only if variable size dist
          !           growth coefficent Abdul-Razzak & Ghan 1998 eqn 16
          !           should depend on mean radius of mode to account for gas kinetic effects
          !           see Fountoukis and Nenes, JGR2005 and Meskhidze et al., JGR2006
          !           for approriate size to use for effective diffusivity.
-         g=1._r8/(rhoh2o/(diff0*rhoair*qs)                                    &
-            +latvap*rhoh2o/(conduct0*tair)*(latvap/(rh2o*tair)-1._r8))
-         sqrtg(m)=sqrt(g)
-         beta=2._r8*pi*rhoh2o*g*gamma
-         etafactor2(m)=1._r8/(na(m)*beta*sqrtg(m))
-         if(hygro(m).gt.1.e-10_r8)then
-            smc(m)=2._r8*aten*sqrt(aten/(27._r8*hygro(m)*amcube(m))) ! only if variable size dist
+         beta=2._r8*pi*rhoh2o*gthermfac*gamma
+         etafactor2(imode)=1._r8/(na(imode)*beta*sqrt(gthermfac))
+         if(hygro(imode) > 1.e-10_r8)then
+            smc(imode)=2._r8*aten*sqrt(aten/(27._r8*hygro(imode)*amcube(imode))) ! only if variable size dist
          else
-            smc(m)=100._r8
+            smc(imode)=100._r8
          endif
          !	    write(iulog,*)'sm,hygro,amcube=',smcrit(m),hygro(m),amcube(m)
       else
-         g=1._r8/(rhoh2o/(diff0*rhoair*qs)                                    &
-            +latvap*rhoh2o/(conduct0*tair)*(latvap/(rh2o*tair)-1._r8))
-         sqrtg(m)=sqrt(g)
-         smc(m)=1._r8
-         etafactor2(m)=etafactor2max ! this should make eta big if na is very small.
+         smc(imode)=1._r8
+         etafactor2(imode)=etafactor2max ! this should make eta big if na is very small.
       endif
-      lnsm(m)=log(smc(m)) ! only if variable size dist
+      lnsm(imode)=log(smc(imode)) ! only if variable size dist
       !	 write(iulog,'(a,i4,4g12.2)')'m,na,amcube,hygro,sm,lnsm=', &
       !                   m,na(m),amcube(m),hygro(m),sm(m),lnsm(m)
+      eta(imode)=etafactor1*etafactor2(imode)
+      zeta(imode)=twothird*sqrt(alw)*aten/sqrt(gthermfac)
```

Comment:
would you please check if zeta has any mode dependence? If not, we can compute it just once outside this loop.

Reply:
It is true, zeta does not have any mode dependence here.  The only potential issue is that zeta is passed as an nmode-dimension array to maxsat.  But it looks like maxsat is not called from anywhere else.  So options are:  1)  could initialize the array zeta(:) = twothird.... before the loop; 2)  convert zeta to a scalar here and in maxsat, as well as assign its value outside the loop.

Reply:
Let's declare `zeta` as a scalar.

Reply:
OK, will make accordingly here and in maxsat.